In [1]:
import pandas as pd
# The github URL
url ='https://raw.githubusercontent.com/Viswa9959/aim5005/refs/heads/main/Air_Quality%20(1).csv'
# Loading  the CSV file
data = pd.read_csv(url)
# Displaying the first few rows
data.head()

,Unique ID,Indicator ID,Name,Measure,Measure Info,Geo Type Name,Geo Join ID,Geo Place Name,Time Period,Start_Date,Data Value,Message
0,179772,640,Boiler Emissions- Total SO2 Emissions,Number per km2,number,UHF42,409.0,Southeast Queens,2015,01/01/2015,0.3,NaN
1,179785,640,Boiler Emissions- Total SO2 Emissions,Number per km2,number,UHF42,209.0,Bensonhurst - Bay Ridge,2015,01/01/2015,1.2,NaN
2,178540,365,Fine particles (PM 2.5),Mean,mcg/m3,UHF42,209.0,Bensonhurst - Bay Ridge,Annual Average 2012,12/01/2011,8.6,NaN
3,178561,365,Fine particles (PM 2.5),Mean,mcg/m3,UHF42,409.0,Southeast Queens,Annual Average 2012,12/01/2011,8.0,NaN
4,823217,365,Fine particles (PM 2.5),Mean,mcg/m3,UHF42,409.0,Southeast Queens,Summer 2022,06/01/2022,6.1,NaN


## STEP1:DATA PREPROCESSING

In [2]:
# Displaying the shape of the data set
data.shape

(18025, 12)

In [3]:
# Displaying the last 5 rows of the data set
data.tail()

,Unique ID,Indicator ID,Name,Measure,Measure Info,Geo Type Name,Geo Join ID,Geo Place Name,Time Period,Start_Date,Data Value,Message
18020,816914,643,Annual vehicle miles traveled,Million miles,per square mile,CD,503.0,Tottenville and Great Kills (CD3),2019,01/01/2019,12.9,NaN
18021,816913,643,Annual vehicle miles traveled,Million miles,per square mile,CD,503.0,Tottenville and Great Kills (CD3),2010,01/01/2010,14.7,NaN
18022,816872,643,Annual vehicle miles traveled,Million miles,per square mile,UHF42,208.0,Canarsie - Flatlands,2010,01/01/2010,43.4,NaN
18023,816832,643,Annual vehicle miles traveled,Million miles,per square mile,UHF42,407.0,Southwest Queens,2010,01/01/2010,65.8,NaN
18024,151658,643,Annual vehicle miles traveled,Million miles,per square mile,UHF42,408.0,Jamaica,2005,01/01/2005,41.0,NaN


In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18025 entries, 0 to 18024
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unique ID       18025 non-null  int64  
 1   Indicator ID    18025 non-null  int64  
 2   Name            18025 non-null  object 
 3   Measure         18025 non-null  object 
 4   Measure Info    18025 non-null  object 
 5   Geo Type Name   18025 non-null  object 
 6   Geo Join ID     18016 non-null  float64
 7   Geo Place Name  18016 non-null  object 
 8   Time Period     18025 non-null  object 
 9   Start_Date      18025 non-null  object 
 10  Data Value      18025 non-null  float64
 11  Message         0 non-null      float64
dtypes: float64(3), int64(2), object(7)
memory usage: 1.7+ MB


## Explanation:
* The DataFrame has 18,025 rows and 12 columns.
* Each column's name, the count of non-null values, and their data types are listed.
* Most columns have 18,025 non-null entries, except Geo Join ID and Geo Place Name, which have 18,016 non-null values, indicating 9 missing values.
* The Message column has 0 non-null values, means it is entirely empty.

In [5]:
data.isnull().sum()

Unique ID             0
Indicator ID          0
Name                  0
Measure               0
Measure Info          0
Geo Type Name         0
Geo Join ID           9
Geo Place Name        9
Time Period           0
Start_Date            0
Data Value            0
Message           18025
dtype: int64

* As pe the above data analysis we can drop the UniqueId,IndicatorId,and Message attributes because these  are ID's and and empty attributes, So they are not useful for the prediction of Air quality.

In [6]:
# Listing the columns to drop
columns_to_drop = ['Unique ID', 'Indicator ID', 'Message']

# Droping the columns from the DataFrame
air_quality_data= data.drop(columns=columns_to_drop)

# Verifying the data frame
air_quality_data.head()


,Name,Measure,Measure Info,Geo Type Name,Geo Join ID,Geo Place Name,Time Period,Start_Date,Data Value
0,Boiler Emissions- Total SO2 Emissions,Number per km2,number,UHF42,409.0,Southeast Queens,2015,01/01/2015,0.3
1,Boiler Emissions- Total SO2 Emissions,Number per km2,number,UHF42,209.0,Bensonhurst - Bay Ridge,2015,01/01/2015,1.2
2,Fine particles (PM 2.5),Mean,mcg/m3,UHF42,209.0,Bensonhurst - Bay Ridge,Annual Average 2012,12/01/2011,8.6
3,Fine particles (PM 2.5),Mean,mcg/m3,UHF42,409.0,Southeast Queens,Annual Average 2012,12/01/2011,8.0
4,Fine particles (PM 2.5),Mean,mcg/m3,UHF42,409.0,Southeast Queens,Summer 2022,06/01/2022,6.1


In [7]:
air_quality_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18025 entries, 0 to 18024
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Name            18025 non-null  object 
 1   Measure         18025 non-null  object 
 2   Measure Info    18025 non-null  object 
 3   Geo Type Name   18025 non-null  object 
 4   Geo Join ID     18016 non-null  float64
 5   Geo Place Name  18016 non-null  object 
 6   Time Period     18025 non-null  object 
 7   Start_Date      18025 non-null  object 
 8   Data Value      18025 non-null  float64
dtypes: float64(2), object(7)
memory usage: 1.2+ MB


__As per the above information we can drop the 9 "Geo place Name" records, Because it has null values and also we cannot able to impute some predicted values also because it is a categorical data__

In [8]:
air_quality_data.dropna(inplace=True)

In [9]:
air_quality_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 18016 entries, 0 to 18024
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Name            18016 non-null  object 
 1   Measure         18016 non-null  object 
 2   Measure Info    18016 non-null  object 
 3   Geo Type Name   18016 non-null  object 
 4   Geo Join ID     18016 non-null  float64
 5   Geo Place Name  18016 non-null  object 
 6   Time Period     18016 non-null  object 
 7   Start_Date      18016 non-null  object 
 8   Data Value      18016 non-null  float64
dtypes: float64(2), object(7)
memory usage: 1.4+ MB


In [10]:
# Displaing the corrected data frame and it is fully cleaned.
air_quality_data.head()

,Name,Measure,Measure Info,Geo Type Name,Geo Join ID,Geo Place Name,Time Period,Start_Date,Data Value
0,Boiler Emissions- Total SO2 Emissions,Number per km2,number,UHF42,409.0,Southeast Queens,2015,01/01/2015,0.3
1,Boiler Emissions- Total SO2 Emissions,Number per km2,number,UHF42,209.0,Bensonhurst - Bay Ridge,2015,01/01/2015,1.2
2,Fine particles (PM 2.5),Mean,mcg/m3,UHF42,209.0,Bensonhurst - Bay Ridge,Annual Average 2012,12/01/2011,8.6
3,Fine particles (PM 2.5),Mean,mcg/m3,UHF42,409.0,Southeast Queens,Annual Average 2012,12/01/2011,8.0
4,Fine particles (PM 2.5),Mean,mcg/m3,UHF42,409.0,Southeast Queens,Summer 2022,06/01/2022,6.1


In [11]:
# Convert the 'Start_Date' column to datetime format.
# This allows us to extract time-based features like year.
air_quality_data['Start_Date'] = pd.to_datetime(air_quality_data['Start_Date'], errors='coerce')

# Extract the 'Year' from 'Start_Date' to use as a feature.
air_quality_data['Year'] = air_quality_data['Start_Date'].dt.year

# Drop rows where 'Start_Date' could not be parsed to avoid errors.
air_quality_data = air_quality_data.dropna(subset=['Start_Date'])

In [12]:
air_quality_data.head()

,Name,Measure,Measure Info,Geo Type Name,Geo Join ID,Geo Place Name,Time Period,Start_Date,Data Value,Year
0,Boiler Emissions- Total SO2 Emissions,Number per km2,number,UHF42,409.0,Southeast Queens,2015,2015-01-01,0.3,2015
1,Boiler Emissions- Total SO2 Emissions,Number per km2,number,UHF42,209.0,Bensonhurst - Bay Ridge,2015,2015-01-01,1.2,2015
2,Fine particles (PM 2.5),Mean,mcg/m3,UHF42,209.0,Bensonhurst - Bay Ridge,Annual Average 2012,2011-12-01,8.6,2011
3,Fine particles (PM 2.5),Mean,mcg/m3,UHF42,409.0,Southeast Queens,Annual Average 2012,2011-12-01,8.0,2011
4,Fine particles (PM 2.5),Mean,mcg/m3,UHF42,409.0,Southeast Queens,Summer 2022,2022-06-01,6.1,2022


In [13]:
air_quality_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 18016 entries, 0 to 18024
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Name            18016 non-null  object        
 1   Measure         18016 non-null  object        
 2   Measure Info    18016 non-null  object        
 3   Geo Type Name   18016 non-null  object        
 4   Geo Join ID     18016 non-null  float64       
 5   Geo Place Name  18016 non-null  object        
 6   Time Period     18016 non-null  object        
 7   Start_Date      18016 non-null  datetime64[ns]
 8   Data Value      18016 non-null  float64       
 9   Year            18016 non-null  int32         
dtypes: datetime64[ns](1), float64(2), int32(1), object(6)
memory usage: 1.4+ MB


## STEP2:FEATURE SELECTION
* Feature selection is the process of choosing a subset of relevant features (variables or predictors) for building a machine learning model. It aims to select the most important features that contribute to the prediction of the target variable while reducing the complexity of the model, enhancing its interpretability, and potentially improving its performance.

## 1st Model Using RandomForest Regressor

In [14]:
# Importing the neccessary libraries 
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

In [15]:
# Defining the features (input variables) and the target (output variable).
# I am going to select 'Geo Place Name', 'Geo Join ID', 'Year', and 'Name' are selected as features.
features = ['Geo Place Name', 'Geo Join ID', 'Year', 'Name']
X = air_quality_data[features] 
y = air_quality_data['Data Value']
X.head()

,Geo Place Name,Geo Join ID,Year,Name
0,Southeast Queens,409.0,2015,Boiler Emissions- Total SO2 Emissions
1,Bensonhurst - Bay Ridge,209.0,2015,Boiler Emissions- Total SO2 Emissions
2,Bensonhurst - Bay Ridge,209.0,2011,Fine particles (PM 2.5)
3,Southeast Queens,409.0,2011,Fine particles (PM 2.5)
4,Southeast Queens,409.0,2022,Fine particles (PM 2.5)


In [16]:
y.head()

0    0.3
1    1.2
2    8.6
3    8.0
4    6.1
Name: Data Value, dtype: float64

## STEP3:One Hot Encoding for Categorical Data

In [17]:
# Many machine learning models cannot work directly with categorical data.
# We need to convert 'Geo Place Name' and 'Name' into numerical values using one-hot encoding.
column_transformer = ColumnTransformer(
    transformers=[
        ('encoder', OneHotEncoder(handle_unknown='ignore'), ['Geo Place Name', 'Name'])
    ],
    remainder='passthrough'
)

## STEP4:Splitting the Data:

* Divide the data into training and testing sets to evaluate the model's performance.

In [18]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## STEP5: Model Selection and Training

In [19]:
# We use RandomForestRegressor, which is robust and handles complex data also.
# A Random Forest consists of many decision trees and averages their predictions.
model = RandomForestRegressor(n_estimators=100, random_state=42)

# A pipeline ensures that data transformation and model training happen together.
pipeline = Pipeline(steps=[
    ('preprocessor', column_transformer),
    ('regressor', model)
])

# Fit the model
pipeline.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('encoder',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['Geo Place Name',
                                                   'Name'])])),
                ('regressor', RandomForestRegressor(random_state=42))])

## STEP6: Model Evaluation

In [20]:
# Use the trained model to predict air quality values for the test set.
y_pred = pipeline.predict(X_test)

# Calculating the model's performance using different metrics:
# - Mean Absolute Error (MAE): Average absolute difference between actual and predicted values.
# - Mean Squared Error (MSE): Average squared difference between actual and predicted values.
# - R-squared (R2): Indicates how well the model explains the variability in the data.
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Absolute Error (MAE): {mae}")
print(f"Mean Squared Error (MSE): {mse}")
print(f"R-squared (R2): {r2}")


Mean Absolute Error (MAE): 4.589237171539279
Mean Squared Error (MSE): 123.16861721691984
R-squared (R2): 0.7616804137935664


__Mean Absolute Error__
* The formula for Mean Absolute Error (MAE) is:

$$
\text{MAE} = \frac{1}{n} \sum|y_t - y_p|
$$

Where:
- \( n \) = number of samples
- \( y_t \) = actual value
- \( y_p) = predicted value

_MAE measures the average magnitude of the errors in a set of predictions, without considering their direction (i.e., absolute difference)._

_In this case, MAE = 4.589, which means that, on average, the model's predictions are off by about 4.59 units from the actual values._

__Mean Squared Error__
* The formula for Mean Suared Error (MSE) is:

$$
\text{MSE} = \frac{1}{n} \sum(y_t - y_p)^2
$$

Where:
- \( n \) = number of samples
- \( y_t \) = actual value
- \( y_p) = predicted value

_MSE measures the average of the squares of the errors, penalizing larger errors more than MAE does._

_Here, MSE = 123.17, indicating the average squared error between the predictions and the actual values._

### R-squared (R²)
The formula for R-squared is:

$$
R^2 = 1 - \frac{\sum (y_{\text{true}} - y_{\text{pred}})^2}{\sum (y_{\text{true}} - \bar{y})^2}
$$

* _R², or the coefficient of determination, measures how well the model explains the variance in the actual data._
* __The R² value ranges from 0 to 1:__
 * Closer to 1: The model explains most of the variance in the target variable.
 * Closer to 0: The model does not explain the variance well.
* Here, R² = 0.7616, which means the model give accuracy of 76.16% of the variability in the air quality values in the test set.

## Output Interpretation For RandomForest Regressor:

__MAE (Mean Absolute Error): 4.589__
* This tells us that, on average, the model's predictions are about 4.59 units off from the actual air quality values. Lower MAE values are better, as they indicate closer predictions.

__MSE (Mean Squared Error): 123.168__
* The MSE is relatively higher compared to MAE due to squaring the errors, which gives more weight to larger errors. While this metric can provide insights into prediction accuracy, it is less interpretable than MAE due to its squared units.

__R² (R-squared): 0.7616__
* The R² value of 0.7616 indicates that the model captures about 76.16% of the variance in the target variable (air quality values) with the input features. It shows that the model performs good.

## Step7: Making Predictions

In [47]:
# Takig the user input  for which year we need the air quality prediction value 
geo_place_name = input("Enter the Geo Place Name (e.g., 'Jamaica'): ")
geo_join_id = int(input("Enter the Geo Join ID (e.g., 150): "))
year = int(input("Enter the Year for which year we need for prediction values (e.g., 2025): "))
pollutant_name = input("Enter the Pollutant Name (e.g., 'Nitrogen dioxide (NO2)'): ")
new_data = pd.DataFrame({
    'Geo Place Name': [geo_place_name],
    'Geo Join ID': [geo_join_id],
    'Year': [year],
    'Name': [pollutant_name]
})

# Use the trained pipeline to predict the air quality value for this new data
predicted_value = pipeline.predict(new_data)
print(f"Predicted Air Quality Value for the year {year} in {geo_place_name} is : {predicted_value[0]}")


Enter the Geo Place Name (e.g., 'Jamaica'):  Jamaica
Enter the Geo Join ID (e.g., 150):  150
Enter the Year for which year we need for prediction values (e.g., 2025):  2025
Enter the Pollutant Name (e.g., 'Nitrogen dioxide (NO2)'):  Nitrogen dioxide (NO2)


Predicted Air Quality Value for the year 2025 in Jamaica is : 20.545333498953724


In [48]:
new_data = pd.DataFrame({
    'Geo Place Name': [geo_place_name],
    'Geo Join ID': [geo_join_id],
    'Year': [year],
    'Name': [pollutant_name],
    'Predictied AQI': [predicted_value[0]]
})
new_data

,Geo Place Name,Geo Join ID,Year,Name,Predictied AQI
0,Jamaica,150,2025,Nitrogen dioxide (NO2),20.545333


# 2nd Model using " GradientBoostingRegressor"

In [51]:
from sklearn.ensemble import GradientBoostingRegressor
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
# Using Gradient Boosting Regressor without hyperparameter tuning
model = GradientBoostingRegressor(random_state=42)
# Combine preprocessing and model into a pipeline
pipeline = Pipeline(steps=[
    ('preprocessor', column_transformer),
    ('regressor', model)
])

# Train the model using the training data
pipeline.fit(X_train, y_train)

# Model Evaluation
# Predict air quality values for the test set
y_pred = pipeline.predict(X_test)

# Calculate the model's performance using different metrics
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Absolute Error (MAE): {mae}")
print(f"Mean Squared Error (MSE): {mse}")
print(f"R-squared (R2): {r2}")
# Takig the user input  for which year we need the air quality prediction value 
geo_place_name = input("Enter the Geo Place Name (e.g., 'Jamaica'): ")
geo_join_id = int(input("Enter the Geo Join ID (e.g., 150): "))
year = int(input("Enter the Year for which year we need for prediction values (e.g., 2025): "))
pollutant_name = input("Enter the Pollutant Name (e.g., 'Nitrogen dioxide (NO2)'): ")
new_data1 = pd.DataFrame({
    'Geo Place Name': [geo_place_name],
    'Geo Join ID': [geo_join_id],
    'Year': [year],
    'Name': [pollutant_name]
})

# Use the trained pipeline to predict the air quality value for this new data
predicted_value1 = pipeline.predict(new_data1)
print(f"Predicted Air Quality Value for the year {year} in {geo_place_name} is : {predicted_value1[0]}")


Mean Absolute Error (MAE): 5.93216333090172
Mean Squared Error (MSE): 173.7107061983137
R-squared (R2): 0.6638862678152849


Enter the Geo Place Name (e.g., 'Jamaica'):  Jamaica
Enter the Geo Join ID (e.g., 150):  340
Enter the Year for which year we need for prediction values (e.g., 2025):  2024
Enter the Pollutant Name (e.g., 'Nitrogen dioxide (NO2)'):  Nitrogen dioxide (NO2)


Predicted Air Quality Value for the year 2024 in Jamaica is : 18.081698044241733


In [52]:
new_data1 = pd.DataFrame({
    'Geo Place Name': [geo_place_name],
    'Geo Join ID': [geo_join_id],
    'Year': [year],
    'Name': [pollutant_name],
    'Predictied AQI': [predicted_value1[0]]
})
new_data1

,Geo Place Name,Geo Join ID,Year,Name,Predictied AQI
0,Jamaica,340,2024,Nitrogen dioxide (NO2),18.081698


## Output Interpretation For Gradientboosting Regressor:

__MAE (Mean Absolute Error): 5.9321__
* This tells us that, on average, the model's predictions are about 5.9321 units off from the actual air quality values. Lower MAE values are better, as they indicate closer predictions.

__Mean Squared Error (MSE): 173.7107061983137__

* MSE is the average of the squared differences between actual and predicted values.
* MSE penalizes larger errors more heavily than MAE because it squares the errors.
* An MSE of 173.71 suggests some larger deviations in predictions, but MSE is harder to interpret in terms of units because it involves squared errors.
R-squared (R²): 0.6638862678152849

__R²: 0.6638862678152849__ 

* It is a measure of how well the model explains the variability in the target variable (air quality value).
* An R² value of 0.66 means that the model explains about 66.4% of the variance in the air quality values based on the input features.
* R² values closer to 1 indicate a better fit, so an R² of 0.66 is moderate, indicating that the model captures some of the trends but could be improved for better predictions.